Code for infeasibility diagnostics called mis (#3172)
Pyomo/pyomo: doc/OnlineDocs/contributed_packages/iis.rst

Infeasibility Diagnostics
!!!!!!!!!!!!!!!!!!!!!!!!!

There are two closely related tools for infeasibility diagnosis:

  - :ref:`Infeasible Irreducible System (IIS) Tool`
  - :ref:`Minimal Intractable System finder (MIS) Tool`

The first simply provides a conduit for solvers that compute an
infeasible irreducible system (e.g., Cplex, Gurobi, or Xpress).  The
second provides similar functionality, but uses the ``mis`` package
contributed to Pyomo.

Infeasible Irreducible System (IIS) Tool
========================================

.. automodule:: pyomo.contrib.iis.iis

.. autofunction:: pyomo.contrib.iis.write_iis

Minimal Intractable System finder (MIS) Tool
============================================

The file ``mis.py`` finds sets of actions that each, independently,
would result in feasibility.  The zero-tolerance is whatever the
solver uses, so users may want to post-process output if it is going
to be used for analysis. It also computes a minimal intractable system
(which is not guaranteed to be unique).  It was written by Ben Knueven

> as part of the watertap project and is governed by a license shown

Reviewer: It may be worth putting a link to WaterTAP here so people don't have to guess what it is!

This part of the report is for nonlinear programs (NLPs).

When we’re trying to reduce the constraint set, for an NLP there may be constraints that when missing cause the solver 
to fail in some catastrophic fashion. In this implementation this is interpreted as failing to get a `results` 
object back from the call to `solve`. In these cases we keep the constraint in the problem but it’s in the 
set of “guard” constraints – we can’t really be sure they’re a source of infeasibility or not, 
just that “bad things” happen when they’re not included.

Perhaps ideally we would put a constraint in the “guard” set if IPopt failed to converge, and only put it in the 
MIS if IPopt converged to a point of local infeasibility. However, right now the code generally makes the 
assumption that if IPopt fails to converge the subproblem is infeasible, though obviously that is far from the truth. 
Hence for difficult NLPs even the “Phase 1” may “fail” – in that when finished the subproblem containing just the 
constraints in the elastic filter may be feasible -- because IPopt failed to converge and we assumed that meant the 
subproblem was not feasible.

> Dealing with NLPs is far from clean, but that doesn’t mean the tool can’t return useful results even when it’s assumptions are not satisfied.

Reviewer: ```suggestion
Dealing with NLPs is far from clean, but that doesn’t mean the tool can’t return useful results even when its assumptions are not satisfied.
```
